### PR TITLE
Allow setting variables for intermediateType in AggregateFunctionSignature

### LINKED ISSUE #9105: Allow setting variables for intermediateType in AggregateFunctionSignature
### Description

In some cases, variables are used in the intermediate type of an aggregate function. For example, as discussed in https://github.com/facebookincubator/velox/discussions/9048, the intermediate type of Spark decimal average aggregate function is `ROW(DECIMAL(min(38, p + 10), s), BIGINT)`, where `p` and `s` are the decimal precision and scale of input value. We need to define the aggregate function signature just like this.
```cpp
exec::AggregateFunctionSignatureBuilder()
    .integerVariable("a_precision")
    .integerVariable("a_scale")
    .integerVariable("i_precision", "min(38, a_precision + 10)")
    .integerVariable("r_precision", "min(38, a_precision + 4)")
    .integerVariable("r_scale", "min(38, a_scale + 4)")
    .argumentType("DECIMAL(a_precision, a_scale)")
    .intermediateType("ROW(DECIMAL(i_precision, a_scale), bigint)")
    .returnType("DECIMAL(r_precision, r_scale)")
    .build()
```
However, Velox does not support having new variables in intermediate types. Declaring `i_precision` will cause an exception to be thrown during `FunctionSignature` validation.
```
Reason: (4 vs. 5) Some integer variables are not used
Retriable: False
Expression: usedVariables.size() == variables.size()
Function: validate
File: ../velox/expression/FunctionSignature.cpp
Line: 168
```

The reason for the error here is that `AggregateFunctionSignature` holds `intermediateType_` but does not override its own validation logic, causing an issue with the check for the number of used variables.

We need to collect used variables in `intermediateType_`, and pass it to the validate function of `FunctionSignature`.

### PR DESCRIPTION
**Pull request by liujiayi771**
Fixes #9105

### CODE REVIEW

File: velox/expression/FunctionSignature.h [hunk 1]
```diff
@@ -109,6 +109,8 @@ class SignatureVariable {
 
 class FunctionSignature {
  public:
+  /// Construct a function signature. Need to call the validate method to
```
Comment (jinchengchenghh) [inline velox/expression/FunctionSignature.h:112]: Does all the constructor should follow validate method? If so, I would suggest to add a new function `construct` to call constructor and validate

Comment (liujiayi771) [reply velox/expression/FunctionSignature.h:112]: virtual method cannot be used in constructor.

Comment (liujiayi771) [reply velox/expression/FunctionSignature.h:112]: We have `FunctionSignatureBuilder` and `AggregateFunctionSignatureBuilder`, we can construct `FunctionSignature` solely through their build methods. This way, we only need to ensure that the `validate` method is called within the build method. I will modify the code in `AggregateCompanionSignatures.cpp` that directly calls the constructor to use the build method for construction instead.

**Comment by liujiayi771** [comment]
Hi @mbasmanova @kagamiori. Could you help review?

**Comment by mbasmanova** [comment]
@liujiayi771 CI is red.

File: velox/expression/tests/FunctionSignatureTest.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ */
+
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox;
+
+class FunctionSignatureTest : public functions::test::FunctionBaseTest {};
+
+TEST_F(FunctionSignatureTest, validateFunctionSignature) {
```
Comment (mbasmanova) [inline velox/expression/tests/FunctionSignatureTest.cpp:24]: There is an existing velox/exec/tests/FunctionSignatureBuilderTest.cpp test. Let's move this new test there.

File: velox/expression/tests/CMakeLists.txt [hunk 3]
```diff
@@ -17,40 +17,41 @@ add_subdirectory(utils)
 add_executable(
   velox_expression_test
   ArgumentTypeFuzzerTest.cpp
+  ArrayViewTest.cpp
```
Comment (mbasmanova) [inline velox/expression/tests/CMakeLists.txt:20]: Thank you for fixing the order of the files here. Would you extract this change into a separate PR?

**Comment by liujiayi771** [comment]
> @liujiayi771 CI is red.
I see. But I find that many other PRs failed, because of these two tests:
* CustomTypeTest.getCustomTypeNames
* GenericViewTest.toString

File: velox/expression/tests/CMakeLists.txt (hunk 3 above)
Comment (liujiayi771) [reply velox/expression/tests/CMakeLists.txt:20]: Alright. I will create a new PR, and move `FunctionSignatureTest` to `FunctionSignatureBuilderTest `, thanks.

File: velox/expression/FunctionSignature.h [hunk 4]
```diff
@@ -171,7 +178,6 @@ class FunctionSignature {
   // Return a string of the list of argument types.
   std::string argumentsToString() const;
 
- private:
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.h:174]: Accidental change?

Comment (liujiayi771) [reply velox/expression/FunctionSignature.h:174]: `AggregateFunctionsSignature` needs to get the `variables_`, so I changed it to `protected`.

File: velox/expression/FunctionSignature.cpp [hunk 5]
```diff
@@ -222,14 +215,23 @@ std::string AggregateFunctionSignature::toString() const {
...
 FunctionSignaturePtr FunctionSignatureBuilder::build() {
   VELOX_CHECK(returnType_.has_value());
-  return std::make_shared<FunctionSignature>(
+  auto funcSignature = std::make_shared<FunctionSignature>(
       std::move(variables_),
       returnType_.value(),
       std::move(argumentTypes_),
       std::move(constantArguments_),
       variableArity_);
+  funcSignature->validate();
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.cpp:233]: Validation logic needs to stay in the constructor of the FunctionSignature. FunctionSignatureBuilder is only one way to create FunctionSignature. Is it also possible to use FunctionSignature directly. We need to make sure FunctionSignature is always valid no matter how it was created.

**Comment by mbasmanova** [comment]
> But I find that many other PRs failed, because of these two tests in velox_expression_test:

Would you try rebasing? If tests still fail, please, create GitHub issue. We'll need to fix these tests before merging PRs.

File: velox/expression/FunctionSignature.cpp (hunk 5 above)
Comment (liujiayi771) [reply velox/expression/FunctionSignature.cpp:233]: @mbasmanova If validation logic is in the constructor, the `FunctionSignature` cannot validate `variables_` because it contains variables used in the `intermediateType_` of `AggregateFunctionSignature`, and the `FunctionSignature` cannot determine whether they are valid.

**Comment by liujiayi771** [comment]
@mbasmanova I have already moved the validation into the constructor, and the `AggregateFunctionSignature` passes the `usedVariables` to `FunctionSignature` for validation through a lambda function.

File: velox/expression/FunctionSignature.h [hunk 6]
```diff
@@ -168,6 +168,16 @@ class FunctionSignature {
   }
 
  protected:
+  FunctionSignature(
+      std::unordered_map<std::string, SignatureVariable> variables,
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.h:172]: pass 'variables' as const &

File: velox/expression/FunctionSignature.h [hunk 7]
```diff
@@ -168,6 +168,16 @@ class FunctionSignature {
...
+      TypeSignature returnType,
+      std::vector<TypeSignature> argumentTypes,
+      std::vector<bool> constantArguments,
+      bool variableArity,
+      const std::function<std::unordered_set<std::string>(
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.h:177]: Would you document this parameter?

File: velox/expression/FunctionSignature.cpp [hunk 8]
```diff
@@ -145,8 +145,8 @@ void validate(
     const std::unordered_map<std::string, SignatureVariable>& variables,
     const TypeSignature& returnType,
     const std::vector<TypeSignature>& argumentTypes,
-    const std::vector<bool>& constantArguments) {
-  std::unordered_set<std::string> usedVariables;
+    const std::vector<bool>& constantArguments,
+    std::unordered_set<std::string> usedVariables) {
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.cpp:149]: Is this an input parameter? Pass by const &

File: velox/expression/FunctionSignature.cpp [hunk 9]
```diff
@@ -212,9 +212,54 @@ FunctionSignature::FunctionSignature(
...
+    bool variableArity)
+    : FunctionSignature(
+          std::move(variables),
+          std::move(returnType),
+          std::move(argumentTypes),
+          std::move(constantArguments),
+          variableArity,
+          [&intermediateType](
+              const std::unordered_map<std::string, SignatureVariable>&
+                  typeVariables) {
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.cpp:255]: I'm not following this design. Would you clarify?

**Review by mbasmanova — COMMENTED** [review]
@liujiayi771 Thank you for iterating. I'm not sure I understand this change. Would you explain?

File: velox/expression/FunctionSignature.cpp (hunk 9 above)
Comment (liujiayi771) [reply velox/expression/FunctionSignature.cpp:255]: @mbasmanova When validating a `FunctionSignature`, it will collect all variables used in the "Type"s and compare this `usedVariables` to the size of `variables_`. If an inconsistency is found, indicating that `variables_` contains unused variables, the `FunctionSignature` is considered invalid.

The `AggregateFunctionSignature` inherits from `FunctionSignature` and adds a new "type" `intermediateType_`. If i`ntermediateType_` uses variables, it will be stored in the `variables_` of `FunctionSignature`. However, during validation of `FunctionSignature`, variables used in `intermediateType_` are not collected. This leads to a situation where the size of the collected `usedVariables` during the `FunctionSignature` validation is less than `variables_.size`, resulting in it being deemed invalid.

Therefore, when validating an `AggregateFunctionSignature`, we need to introduce a lambda function. This lambda function will first validate `intermediateType_` and collect its used variables into `usedVariables`. This `usedVariables`, which now includes the variables from `intermediateType_`, will then be passed as an argument to the new constructor of `FunctionSignature` for validation. In this way, `FunctionSignature` receives a `usedVariables` that already contains the variables from `intermediateType_`. Consequently, when checking if `usedVariables.size` matches `variables_.size`, no issues will arise.

File: velox/expression/FunctionSignature.cpp (hunk 8 above)
Comment (liujiayi771) [reply velox/expression/FunctionSignature.cpp:149]: `validateBaseTypeAndCollectTypeParams` will add elements to this set, so it cannot be const &.

File: velox/expression/FunctionSignature.h (hunk 6 above)
Comment (liujiayi771) [reply velox/expression/FunctionSignature.h:172]: The constructor uses `std::move(variables)` to initialize `variables_` , therefore it cannot be const &.

**Comment by liujiayi771** [comment]
Hi @mbasmanova, I added documents to the new parameter, and the issue I hope to solve is being able to use signature variables in `AggregateFunctionSignature`'s `intermediateType_`. This might be described in more detail in the [issue](https://github.com/facebookincubator/velox/issues/9105). The primary role of the lambda function is to pass the variables used in `intermediateType_` to the constructor of `FunctionSignature` for use during validation.

File: velox/expression/FunctionSignature.cpp (hunk 9 above)
Comment (mbasmanova) [reply velox/expression/FunctionSignature.cpp:255]: @liujiayi771 Thank you for clarifying. 

> If the subclass introduces a new Type variable and that Type uses signature variables

If that's the only use case, then maybe a simple solution is to pass a list of additional types to the base constructor for validation. What do you think?

Comment (liujiayi771) [reply velox/expression/FunctionSignature.cpp:255]: @mbasmanova We need to use `validateBaseTypeAndCollectTypeParams` to collect the used variables in `intermediateType` first, and then pass it to the constructor of `FunctionSignature`. Can we achieve this without using lambda functions?

Comment (liujiayi771) [reply velox/expression/FunctionSignature.cpp:255]: @mbasmanova I understand what you mean. If it's about passing the type, I think this way is simpler.

File: velox/expression/FunctionSignature.cpp [hunk 10]
```diff
@@ -212,7 +212,32 @@ FunctionSignature::FunctionSignature(
...
+    std::vector<bool> constantArguments,
+    bool variableArity,
+    const std::vector<TypeSignature>& additionalTypes)
+    : variables_{std::move(variables)},
+      returnType_{std::move(returnType)},
+      argumentTypes_{std::move(argumentTypes)},
+      constantArguments_{std::move(constantArguments)},
+      variableArity_{variableArity} {
+  std::unordered_set<std::string> usedVariables;
+  for (const auto& type : additionalTypes) {
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.cpp:231]: perhaps, move this logic into 'validate'

**Review by mbasmanova — COMMENTED** [review]
@liujiayi771 Looks great % one nit.

**Review by mbasmanova — APPROVED** [review]
@liujiayi771 Looks great. Thank you for iterating to close this gap.

**Comment by mbasmanova** [comment]
@liujiayi771 Would you make sure the PR is rebased onto latest main?

**Comment by liujiayi771** [comment]
> @liujiayi771 Would you make sure the PR is rebased onto latest main?

Rebased.
